File: hard/4_smallest_difference/smallest_difference.js
function smallestDifference(array1, array2) {
	array1.sort((a, b) => a - b);
	array2.sort((a, b) => a - b);
	let idxOne = 0;
	let idxTwo = 0;
	let smallest = Number.MAX_SAFE_INTEGER;
	let current = Number.MAX_SAFE_INTEGER;
	let smallestPair = [];
	while (idxOne < array1.length && idxTwo < array2.length) {
		const first = array1[idxOne];
		const second = array2[idxTwo];
		if (first < second) {
			current = second - first;
			idxOne += 1;
		} else if (second < first) {
			current = first - second;
			idxTwo += 1;
		} else {
			return [first, second];
		}
		if (smallest > current) {
			smallest = current;
			smallestPair = [first, second];
		}
	}
	return smallestPair;
}

const array1 = [-1, 5, 10, 20, 28, 3];
const array2 = [26, 134, 135, 15, 17];
console.log(smallestDifference(array1, array2));

module.exports = smallestDifference;

File: hard/4_smallest_difference/smallest_difference.py
from typing import List


def smallest_difference(array1: List[int], array2: List[int]) -> List[int]:
    array1.sort()
    array2.sort()
    idx_one = 0
    idx_two = 0
    smallest = float("inf")
    current = float("inf")
    smallest_pair = []
    while idx_one < len(array1) and idx_two < len(array2):
        first = array1[idx_one]
        second = array2[idx_two]
        if first < second:
            current = second - first
            idx_one += 1
        elif second < first:
            current = first - second
            idx_two += 1
        else:
            return [first, second]
        if smallest > current:
            smallest = current
            smallest_pair = [first, second]
    return smallest_pair


if __name__ == "__main__":
    array1 = [-1, 5, 10, 20, 28, 3]
    array2 = [26, 134, 135, 15, 17]
    print(smallest_difference(array1, array2))